core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala

```diff
@@ -0,0 +1,474 @@
+    idVerifierRef.ask[Boolean](ID(endpointRef.name)).flatMap(find =>
+      if (find) {
+        Future.successful(endpointRef)
+      } else {
+        Future.failed(new RpcEndpointNotFoundException(uri))
+      }
+    )(ThreadUtils.sameThread)
+  }
+
+  override def stop(endpointRef: RpcEndpointRef): Unit = {
+    require(endpointRef.isInstanceOf[NettyRpcEndpointRef])
+    dispatcher.stop(endpointRef)
+  }
+
+  private[netty] def send(message: RequestMessage): Unit = {
+    val remoteAddr = message.receiver.address
+    if (remoteAddr == address) {
+      val promise = Promise[Any]()
+      dispatcher.postMessage(message, promise)
+      promise.future.onComplete {
+        case Success(response) =>
+          val ack = response.asInstanceOf[SendAck]
+          logDebug(s"Receive ack from ${ack.sender}")
+        case Failure(e) =>
+          logError(s"Exception when sending $message", e)
+      }(ThreadUtils.sameThread)
+    } else {
+      try {
+        // `createClient` will block if it cannot find a known connection, so we should run it in
+        // clientConnectionExecutor
+        clientConnectionExecutor.execute(new Runnable {
+          override def run(): Unit = Utils.tryLogNonFatalError {
+            val client = clientFactory.createClient(remoteAddr.host, remoteAddr.port)
+            client.sendRpc(serialize(message), new RpcResponseCallback {
+
+              override def onFailure(e: Throwable): Unit = {
+                logError(s"Exception when sending $message", e)
+              }
+
+              override def onSuccess(response: Array[Byte]): Unit = {
+                val ack = deserialize[SendAck](response)
+                logDebug(s"Receive ack from ${ack.sender}")
+              }
+            })
+          }
+        })
+      } catch {
+        case e: RejectedExecutionException => {
+          // `send` after shutting clientConnectionExecutor down, ignore it
+        }
+      }
+    }
+  }
+
+  private[netty] def ask(message: RequestMessage): Future[Any] = {
+    val promise = Promise[Any]()
+    val remoteAddr = message.receiver.address
+    if (remoteAddr == address) {
+      val p = Promise[Any]()
+      dispatcher.postMessage(message, p)
+      p.future.onComplete {
+        case Success(response) =>
+          val reply = response.asInstanceOf[AskResponse]
+          if (reply.reply != null && reply.reply.isInstanceOf[RpcFailure]) {
+            if (!promise.tryFailure(reply.reply.asInstanceOf[RpcFailure].e)) {
+              logWarning(s"Ignore failure: ${reply.reply}")
+            }
+          } else if (!promise.trySuccess(reply.reply)) {
+            logWarning(s"Ignore message: ${reply}")
+          }
+        case Failure(e) =>
+          if (!promise.tryFailure(e)) {
+            logWarning("Ignore Exception", e)
+          }
+      }(ThreadUtils.sameThread)
+    } else {
+      val client = clientFactory.createClient(remoteAddr.host, remoteAddr.port)
```

Comment:
So, in `send()` you avoid calling this directly since it can block, using `clientConnectionExecutor` to actually send the message. Why isn't that necessary here?